guides/terminology.md

```diff
@@ -0,0 +1,89 @@
+## Terminology
+
+### Mutation Backflow
+
+Mutation backflow occurs when a child context mutates an object that is owned
+by a parent context, causing changes in the parent context.  This is analogous
+to "spooky action at a distance".
+ 
+An example negative side-effect of mutation backflow occurs when a parent
+component has already rendered when a child component mutates an object owned
+or utilized by the parent component.  This forces the parent component to need
+to re-render, and is one of several causes of what we refer to as a backtracking
+rerender.
+
+### Backtracking Rerender
+
+A backtracking rerender occurs when already rendered data is mutated before the
+entire render completes. This forces a second render pass to account for any
+updated state. Backtracking rerenders are synchronous and will block paint.
+
+### The State Tree
+
+In a web application, state can live at many different scopes. We can represent
+the scopes from outermost to innermost as a tree.
+ 
+* Global State (state that is universal, would be true for any number of
+  applications on the same window, such as in SSR or when running tests)
+  * Application State (state that should live and die with the application
+    instance, the application route and services fall into this category)
+    * Route State (state that should live and die based on a route 
+      activating / deactivating)
+      * Local State (state that should live and die based on the display
+        lifecycle of a single component, also known as "UI State" or
+        "Component State")
+ 
+You may notice similarities between the state tree and a DOM Tree. Because the
+application has a root element, and routes nest as children within it, and
+components and their subcomponents as children within routes, the DOM is a
+reflection of the full state of the application.
+
+### Data Down, Actions Up
+
+"Data down, actions up" represents the philosophy through which state in the
+state tree should be managed.
+ 
+"Data Down" because data should be fetched near the top of the tree (at the
+Application or Route level) and fed down to components.
+ 
+"Data Down" also refers to yielding derived state out of a shared parent
+component into multiple child components.
+ 
+"Actions Up" because mutations to data should occur at the same level of the
+state tree that owns or originated that data.  Components further down the tree
+send an action back "up" the tree to the owner of the data signaling the piece
+of data and the requested mutation.  For derived data this may be a parent
+component.  For Application or Route level state the action should be sent to a
+route or service.
+
+### Sideways Communication
```

Comment:
There has to be a better way to describe this. With data down, actions up, don't go sideways -- I feel like I need a compass to build my app.

Let's focus more on example-driven guidance, and challenge ourselves to avoid coining any new terms. The learning curve is plenty steep already, and new jargon is a new barrier to enter.

Reply:
> The learning curve is plenty steep already

Does anybody see this as a fundamental issue?

Just looking through these issues really lays bare how difficult it is to work with ember and reason about executing complex features.  The cognitive load is enormous.

In my experience, isolated examples are not enough.  Negotiating this complexity holistically is what's required to build an app.